Tests/UnitTests/Utilities/ObservableThreadSafetyTests.swift

```diff
@@ -0,0 +1,437 @@
+//
+// Copyright (c) 2025 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+@_spi(AdyenInternal) @testable import Adyen
+import XCTest
+
+/// Tests for thread-safety of the Observable system with NSLock implementation
+/// These tests verify that the Observable's internal locks prevent crashes and data corruption
+class ObservableThreadSafetyTests: XCTestCase, AdyenObserver {
+    
+    // MARK: - Dictionary Corruption Tests
+
+    // These tests would crash or corrupt data if locks weren't working
+    
+    func testConcurrentHandlerAdditionDoesNotCorruptDictionary() {
+        let observable = AdyenObservable("test")
+        let iterations = 1000
+        
+        // Concurrently add handlers - would crash without proper locking
+        DispatchQueue.concurrentPerform(iterations: iterations) { _ in
+            _ = observable.addEventHandler { _ in }
+        }
+        
+        // If we got here without crashing, locks are working
+        // Verify dictionary is in valid state by triggering all handlers
+        var callCount = 0
+        _ = observable.addEventHandler { _ in
+            callCount += 1
+        }
+        
+        observable.wrappedValue = "trigger"
+        
+        // Should have called the new handler plus all the concurrent ones
+        XCTAssertGreaterThan(callCount, 0)
+    }
+    
+    func testConcurrentHandlerRemovalDoesNotCorruptDictionary() {
+        let observable = AdyenObservable("test")
+        let iterations = 1000
+        
+        // Add many handlers
+        var tokens = [EventHandlerToken]()
+        for _ in 0..<iterations {
+            let token = observable.addEventHandler { _ in }
+            tokens.append(token)
+        }
+        
+        // Concurrently remove them - would crash without proper locking
+        DispatchQueue.concurrentPerform(iterations: iterations) { index in
+            observable.removeEventHandler(with: tokens[index])
+        }
+        
+        // Dictionary should be empty or nearly empty
+        observable.wrappedValue = "test"
+        
+        // If we got here without crashing, locks are working
+        XCTAssertTrue(true)
+    }
+    
+    func testConcurrentAddAndRemoveDoesNotCrash() {
+        let observable = AdyenObservable(0)
+        let iterations = 500
+        let expectation = expectation(description: "All operations complete")
+        expectation.expectedFulfillmentCount = iterations * 2
+        
+        // Simultaneously add and remove handlers
+        DispatchQueue.concurrentPerform(iterations: iterations) { _ in
```

Comment:
Just a note - I would make all tests red first to ensure they actually test thread-safety. If tests do not crash with the old code, they are meaningless.